### PR TITLE
FEAT: adjust and test RCNN model

### PR DESCRIPTION
**Pull request by hopperjc**

### CODE REVIEW

File: vespa/methods/detection/rcnn/model.py [hunk 1]
```diff
@@ -18,23 +18,37 @@ def __init__(
...
-        self.model.roi_heads.box_predictor = FastRCNNPredictor(
-            in_features, num_classes
+        super().__init__(*args, **kwargs)
+
+        # Define os pesos corretamente
+        if num_classes != 91:
+            weights = None
+
+        # Verifica se há pesos para o backbone nos argumentos
+        weights_backbone = kwargs.get("weights_backbone", None)
```
Comment (goncamateus) [inline vespa/methods/detection/rcnn/model.py:31]: deixa como argumento no init mesmo pow

```suggestion
        def __init__(
        self,
        num_classes: int = 9,
        weights: Optional[str] = 'DEFAULT',
        optimizer_name: str = 'adam',
        lr: float = 0.0001,
        weight_decay: float = 0.0001,
        weights_backbone: Optional = None,
        *args,
        **kwargs,
    ):
```

File: tests/methods/detection/rcnn/test_model.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
+from torch import Tensor
+
+
+def test_model_list(rcnn_pretrained_fixture, tensor_image_fixture):
+    model = rcnn_pretrained_fixture
```
Comment (goncamateus) [inline tests/methods/detection/rcnn/test_model.py]: Aqui tu num precisa declarar as variaveis "model" e "image", ja usa direto as fixtures.

File: tests/methods/detection/rcnn/test_model.py
Comment (goncamateus) [inline tests/methods/detection/rcnn/test_model.py:1]: Acho que esses testes nao tao testando algo de interessante.
Pode colocar um exemplo de resultado pra testar se vai dar certo, com msm shape etc.
tipo:
torch.seed(0)
input = torch.ones((1, 32, 32, 3))
result = model(input)

File: vespa/datasets/utils.py [hunk 3]
```diff
@@ -1,4 +1,4 @@
-from torch import Tensor, empty, float32, int64, tensor
+from torch import Tensor, empty, float32, int64, tensor, zeros
```
Comment (goncamateus) [inline vespa/datasets/utils.py]: Aqui tu tas usando Tensor e tensor.
Cuidado que os dois tem um resultado parecido, mas algumas vezes, da muito errado (silent error)

File: vespa/datasets/utils.py [hunk 4]
```diff
@@ -58,3 +58,66 @@ def yolo_to_retinanet(idx: int, image: Tensor, boxes: list, labels: list):
...
+            format [xmin, ymin, xmax, ymax].\n
+            - 'labels' (Tensor): Tensor of labels.\n
+            - 'image_id' (Tensor): Tensor containing the image index.\n
+            - 'area' (Tensor): Tensor containing the area of each bounding box.\n
+            - 'iscrowd' (Tensor): Tensor indicating if the instance is a crowd.
+    """
+    converted_boxes = []
+    img_h, img_w = image.shape[1], image.shape[2]  # Altura e largura corretas
+
+    for box in boxes:
```
Comment (goncamateus) [inline vespa/datasets/utils.py]: Esse trecho aqui tem la em cima tb
Transforma em uma funcaozinha

Comment (goncamateus) [reply vespa/datasets/utils.py]: Na real, a unica coisa que muda eh o "iscrowd".
Da pra reaproveitar muita coisa aqui

**Review by goncamateus — CHANGES_REQUESTED** [review]

File: vespa/methods/detection/rcnn/model.py (hunk 1 above)
Comment (hopperjc) [reply vespa/methods/detection/rcnn/model.py:31]: Se eu deixar com os pesos DEFAULT ele define automaticamente o número de classes que serão utilizadas igual a 91 dentro da RCNN.

File: vespa/datasets/utils.py (hunk 3 above)
Comment (hopperjc) [reply vespa/datasets/utils.py]: O Tensor só é utilizado para declarar o tipo da imagem quando a função é chamada. Todas as tranformações são feitas com tensor.

File: tests/methods/detection/rcnn/test_model.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+def test_model_labels(rcnn_pretrained_fixture, tensor_image_fixture):
+    rcnn_pretrained_fixture.eval()
+    results = rcnn_pretrained_fixture(tensor_image_fixture)
+
+    for result in results:
+        assert isinstance(result['labels'], Tensor)
+
+
+def test_model_train_loop_gpu(
```
Comment (goncamateus) [inline tests/methods/detection/rcnn/test_model.py]: Essa nao ta testando nada
Tu pode verificar se os pesos continuam o mesmos, ou se as logprobs estao as mesmas

File: tests/methods/detection/rcnn/test_model.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+    rcnn_pretrained_fixture, yolo_dataset_train_rcnn
+):
+    rcnn_pretrained_fixture.fit(yolo_dataset_train_rcnn, 4, 4, 0)
+
+
+def test_model_train_loop_cpu(
```
Comment (goncamateus) [inline tests/methods/detection/rcnn/test_model.py]: Essa nao ta testando nada
Tu pode verificar se os pesos continuam o mesmos, ou se as logprobs estao as mesmas

File: vespa/datasets/utils.py [hunk 7]
```diff
@@ -1,7 +1,7 @@
-from torch import Tensor, empty, float32, int64, tensor
+from torch import Tensor, empty, float32, int64, tensor, zeros
 
 
-def yolo_to_retinanet(idx: int, image: Tensor, boxes: list, labels: list):
+def yolo_to_model(idx: int, image: Tensor, boxes: list, labels: list, model: str):
```
Comment (goncamateus) [inline vespa/datasets/utils.py]: Aqui o que eu tinha comentado era pra tu fazer tipo:

def yolo_preprocess(): ...
def yolo_to_retinanet(): ...
    yolo_preprocess()
    ...
def yolo_to_rcnn():
    yolo_preprocess()
    ...

**Review by goncamateus — CHANGES_REQUESTED** [review]

File: tests/methods/detection/rcnn/test_model.py [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+
+def test_model_sketch_labels(rcnn_sketch_fixture, tensor_image_fixture):
+    rcnn_sketch_fixture.eval()
+    results = rcnn_sketch_fixture(tensor_image_fixture)
+
+    for result in results:
+        assert isinstance(result['labels'], Tensor)
+
+
+def test_model_sketch_train_loop_gpu(
```
Comment (goncamateus) [inline tests/methods/detection/rcnn/test_model.py:93]: Remove esse teste.
Em pc sem gpu (que eh o caso do github) vai falhar sempre

**Review by goncamateus — CHANGES_REQUESTED** [review]

**Review by goncamateus — APPROVED** [review]
